el-patch-template.el

```diff
@@ -0,0 +1,513 @@
+             (el-patch-template--any-p (cdr definition)
+                                       ptemplates
+                                       (1- matched-forms-count)))))
+      (error "A form matching a template has subforms matching\
+ other templates"))
+     (t
+      ;; The old resolution of the template uniquely matches the definition
+      ;; Here we first mark the template as being matched then
+      ;; do the actual resolution
+      (plist-put matched-ptemplate :matched t)
+      (let ((resolution
+             (el-patch-template--process definition
+                                         (list
+                                          (plist-get matched-ptemplate
+                                                     :template)))))
+        (cons (caar resolution)
+              (el-patch-template--apply (cdr resolution)
+                                        ptemplates)))))))
+
+(defun el-patch-template--resolve (forms)
+  "Calls `el-patch--resolve' with a special treatment for
+`el-patch-concat'. Specifically, if the arguments of
+`el-patch-concat' have `...' in them, it is not resolved but
+changed to `el-patch-template--concat'."
+  (cl-letf* ((old-concat (symbol-function 'concat))
+             ((symbol-function 'concat)
+              (lambda (&rest args)
+                (if (cl-some (lambda (x) (equal x '...)) args)
+                    (cons 'el-patch-template--concat args)
+                  (apply old-concat args)))))
+    (el-patch--resolve forms nil)))
+
+(defun el-patch-template--impl (keyword-name templates)
+  "The actual implementation of `el-patch-template', accepts the
+same arguments but quoted."
+  (let* ((definition (or (el-patch--locate
+                          (car (el-patch--resolve keyword-name nil)))
+                         (error "Cannot find definition for `%s'"
+                                (cadr keyword-name))))
+         (ptemplates (mapcar
+                      (lambda (template)
+                        (list :template template
+                              :old (el-patch-template--resolve template)
+                              :matched nil))
+                      templates))
+         (patch (prog1 (el-patch-template--apply definition ptemplates)
+                  (cl-dolist (ptemplate ptemplates)
+                    (unless (plist-get ptemplate :matched)
+                      (error
+                       "At least one template did not match any form"))))))
+    ;; NOTE: Unfortunately `el-patch-deftype-alist' doesn't save the
+    ;; macro name so we have to assume that it is `el-patch-*'
+    (cons (intern (format "el-patch-%S"
+                          (car patch))) ;; should be an el-patch-*
+          (append (cdr keyword-name)
+                  (cddr patch)))))
+
+(defmacro el-patch-template (keyword-name &rest templates)
+  "KEYWORD-NAME is cons whose car is a type which can be any type
+from `el-patch-deftype-alist' and the cdr is the name of the form
```

Comment:
Is this true? I think it's a list of two elements, rather than a cons. (The latter would be `(defun . name-of-function)`.)

Reply:
Right, corrected.